File: Data Structures & Algorithms/reverse-a-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head: 
            return None
            
        curr = head
        out = []
        while curr:
            out.append(curr.val)
            curr = curr.next
        out.reverse()

        res = ListNode()
        curr = res

        for x in out:
            curr.next = ListNode(val=x)
            curr = curr.next

        return res.next

File: Data Structures & Algorithms/reorder-linked-list/submission-2.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {void}
     */
    reorderList(head) {
        const list = [];
        let curr = head;

        while (curr) {
            list.push(curr);
            curr = curr.next;
        }

        let i = 0;
        let j = list.length - 1;

        while (i < j) {
            list[i].next = list[j];
            i++;

            if (i >= j) {
                break;
            }

            list[j].next = list[i];
            j--;
        }
        list[i].next = null;
    }
}
